modules/data.land/R/soilgrids_texture_extract.R

```diff
@@ -0,0 +1,129 @@
+##' soilgrids_texture_extraction function
+##' A function to extract and save three types of soil texture data in parallel for a single or group of 
+##' lat/long locations based on user-defined site location from SoilGrids250m 
+##' version 2.0 : https://soilgrids.org
+##' @title soilgrids_texture_extraction
+##' @name soilgrids_texture_extraction
+##' 
+##' @param site_info A data frame of site info containing the BETYdb site ID, 
+##' site name, latitude, and longitude, e.g. 
+##' (site_id, site_name, lat, lon)
+##' @param outdir  Provide the path to store the texture data file
+##' @param verbose Provide progress feedback to the terminal? TRUE/FALSE
+##' @return a data frame containing the soil texture data with columns "Depth", "Quantile", "Siteid", and "Value"
+##' 
+##' @export
+##' @author Qianyu Li
+##' @importFrom magrittr %>%
+
+soilgrids_texture_extraction <- function(site_info, outdir=NULL, verbose=TRUE){
+  
+  # A function to extract and save one type of soil texture data 
+  download_and_extraction <- function(base_data_url, site_info, outdir) {
+    
+    if (future::supportsMulticore()) {
+      future::plan(future::multicore)
+    } else {
+      future::plan(future::multisession)
+    }
+    
+    if (is.null(site_info)) {
+      PEcAn.logger::logger.error(
+        "No site information found. Please provide a BETY DB site list containing at least the site id and PostGIS geometry\
+    as lon and lat"
+      )
+    }
+    
+    # Prepare site info for extraction
+    internal_site_info <- data.frame(site_info$site_id, site_info$site_name, site_info$lat,site_info$lon)
+    #Create a variable to store mean and quantile of soil texture data for each soil depth
+    soiltquant <- matrix(NA, nrow = 6, ncol = length(internal_site_info$site_info.lon) * 4)
+    lonlat <-cbind(internal_site_info$site_info.lon, internal_site_info$site_info.lat)
+    depths <-c("0-5cm", "5-15cm", "15-30cm", "30-60cm", "60-100cm", "100-200cm")
+    
+    # Reproject locations to soilgrids projection
+    # Soilgrids data is using Homolosine projection https://www.isric.org/explore/soilgrids/faq-soilgrids
+    p <-terra::vect(lonlat, crs = "+proj=longlat +datum=WGS84") # Users need to provide lon/lat
+    newcrs <- "+proj=igh +datum=WGS84 +no_defs +towgs84=0,0,0"
+    p_reproj <- terra::project(p, newcrs) # Transform the point vector to data with Homolosine projection
+    data_tag <- c("_mean.vrt", "_Q0.05.vrt", "_Q0.5.vrt", "_Q0.95.vrt")
+    name_tag <- expand.grid(depths, data_tag, stringsAsFactors = F) #find the combinations between data and depth tags.
+    L <- split(as.data.frame(name_tag), seq(nrow(as.data.frame(name_tag))))#convert tags into lists.
+    
+    if ("try-error" %in% class(try(soilt_real <- L %>% furrr::future_map(function(l){
```

Comment:
When the site number goes up, this parallel will be more likely to produce errors due to the limitation of the `SoilGrids` server. The extraction speed is similar with or without the parallel in my previous `6400 site runs`. Thus, it's recommended to just delete the parallel extraction part from the script.

Reply:
Fixed.